packages/react-native/ReactCommon/react/performance/timeline/PerformanceObserver.h

```diff
@@ -0,0 +1,125 @@
+
+/**
+ * Represents subset of spec's `PerformanceObserverInit` that is allowed for multiple types.
+ *
+ * https://w3c.github.io/performance-timeline/#performanceobserverinit-dictionary
+ */
+struct PerformanceObserverObserveMultipleOptions {
+  double durationThreshold = 0.0;
+};
+
+/**
+ * Represents subset of spec's `PerformanceObserverInit` that is allowed for single type.
+ *
+ * https://w3c.github.io/performance-timeline/#performanceobserverinit-dictionary
+ */
+struct PerformanceObserverObserveSingleOptions {
+  bool buffered = false;
+  double durationThreshold = 0.0;
+};
+
+/**
+ * Represents native counterpart of performance timeline PerformanceObserver
+ * class. Each instance has its own entry buffer and can listen for different
+ * performance entry types.
+ *
+ * Entries are pushed to the observer by the `PerformanceEntryReporter` class,
+ * through the `PerformanceObserverRegistry` class which acts as a central hub.
+ *
+ * To create new performance observers, you should use
+ * `PerformanceObserverRegistry::createObserver` which creates the observer,
+ * registers it immediately and unregisters it when it goes out of scope.
+ *
+ * You can still manually create instance of `PerformanceObserver`, but then
+ * you need to manually register and unregister it from registry.
+ */
+class PerformanceObserver: public jsi::NativeState {
+ public:
+  using Ptr = std::shared_ptr<PerformanceObserver>;
+  using WeakPtr = std::weak_ptr<PerformanceObserver>;
+
+  explicit PerformanceObserver(PerformanceObserverCallback&& callback)
+      : callback_(std::move(callback)) {}
+
+  virtual ~PerformanceObserver() {}
+
+  /**
+   * Append entry to the buffer if this observer should handle this entry.
+   */
+  void handleEntry(const PerformanceEntry& entry);
+
+  /**
+   * Returns current observer buffer and clears it.
+   *
+   * Spec:
+   *  https://w3c.github.io/performance-timeline/#takerecords-method
+   */
+  [[nodiscard]] std::vector<PerformanceEntry> takeRecords();
+
+  /**
+   * Configures the observer to watch for specified entry type.
+   *
+   * This operation resets and overrides previous configurations. So consecutive
+   * calls to this methods remove any previous watch configuration (as per spec).
+   */
+  void observe(PerformanceEntryType type, PerformanceObserverObserveSingleOptions options = {});
+
+  /**
+   * Configures the observer to watch for specified entry type.
+   *
+   * This operation resets and overrides previous configurations. So consecutive
+   * calls to this methods remove any previous watch configuration (as per spec).
+   */
+  void observe(std::unordered_set<PerformanceEntryType> types, PerformanceObserverObserveMultipleOptions options = {});
+
+  /**
+   * Internal function called by JS bridge to get number of dropped entries count
+   * counted at call time.
+   */
+  double getDroppedEntriesCount() noexcept;
+
+  /**
+   * Called when the callback was dispatched
+   */
+  void flush() noexcept;
```

Comment:
This shouldn't be necessary. We should handle the side effects from this in `takeRecords`.

Reply:
This looks like a good idea, however this would diverge a bit from the spec. When the callback is scheduled (aka a performance entry is queued), calling `takeRecords` would cancel the `callback` (even though it would be passed an empty list). I will check how it works on browser tomorrow.

Reply:
The use case for calling `takeRecords` manually on the observer is to pull pending entries before disconnecting the observer, so that case should probably never happen anyway. Even if that was a divergence with the spec I'd say it'd be an acceptable one.